### DISCUSSION TITLE
HTPPS and HTTP issue

### DISCUSSION THREAD

**Discussion by PrimeTheFirst** [miguelgrinberg/Flask-SocketIO#1977] [Q&A]
`Mixed Content: The page at 'https://.../chat' was loaded over HTTPS, but requested an insecure XMLHttpRequest endpoint 'http://.../socket.io/?EIO=4&transport=polling&t=OUdK2XB'. This request has been blocked; the content must be served over HTTPS. c
`
When I worked locally, my chat app worked perfectly, but I pushed the changes to my site, which is https, it gave the error. I'm not very qualified with networking and sockets and such , so help would be greatly appreciated as soon as possible. Thanks!

**Comment by miguelgrinberg** [comment]
This isn't really a concern of Flask-SocketIO. If you host your app on HTTPS and use https:// URLs to connect to Socket.IO everything should work the same as in your local environment. This error occurs because your front end is using a http:// URL to connect to the Socket.IO server.

**Comment by PrimeTheFirst** [comment]
Oh thanks! That never came to my mind! Thanks again!

**Comment by PrimeTheFirst** [comment]
What about
`Socket connection to 'wss://../
socket.io/?EIO=4&transport=websocket&sid=CQTIKIcJrqgzaCUpAAAA' failed:
WebSocket is closed before the connection is established`.?

**Reply by miguelgrinberg** [reply]
I can't really debug your application for you. If I had to guess I'd say you do not have a working WebSocket set up, but it would be up to you to debug this and figure out why your WebSocket connection is not working. Flask-SocketIO has documentation, which may help.

**Comment by PrimeTheFirst** [comment]
Any difference between prod and dev which might cause this? Because it works locally.

**Comment by PrimeTheFirst** [comment]
What is meant by invalid session? I read https://stackoverflow.com/questions/74596602/flask-socket-invalid-session-lwmxdnrmai59brflaaaa-further-occurrences-of-this. but I'm not using any websocket like gunicorn or eventlet. It's a prod problem I think. Can it possibly be related to the versions?

**Comment by PrimeTheFirst** [comment]
Im getting invalid session

**Reply by miguelgrinberg** [reply]
I can't really suggest anything because you aren't sharing any information besides the error. Please read the documentation to learn how to run Flask-SocketIO, you are likely doing something wrong.

**Reply by PrimeTheFirst** [reply]
I know! It works properly locally! There's no more information to share , except that the url returned invalid session! Everything has been done according to the documentation, and if it had not, it would not have run locally! I have been struggling with this issue for half a year, and nothing works. Forget it, I'll fix it myself

**Reply by bluthen** [reply]
> I know! It works properly locally! There's no more information to share , except that the url returned invalid session! Everything has been done according to the documentation, and if it had not, it would not have run locally! I have been struggling with this issue for half a year, and nothing works. Forget it, I'll fix it myself

Like on prod do you have a reverse proxy running? Do you got any kind of load balancing going? Sounds like you got https and http running on prod. Sounds like there can be a number of differences between prod and dev for you.

**Reply by PrimeTheFirst** [reply]
I made some changes, did some debugging, and got a new error.
Complete error with traceback
`pp 132017 output: [ pid=132017, time=2023-04-24 11:46:24,425 ]: WSGI application raised an exception!
App 132017 output: Traceback (most recent call last):
App 132017 output:   File "/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/wsgi-loader.py", line 199, in main_loop
App 132017 output:     socket_hijacked = self.process_request(env, input_stream, client)
App 132017 output:   File "/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/wsgi-loader.py", line 333, in process_request
App 132017 output:     result = self.app(env, start_response)
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/flask/app.py", line 2548, in __call__
App 132017 output:     return self.wsgi_app(environ, start_response)
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/flask_socketio/__init__.py", line 43, in __call__
App 132017 output:     return super(_SocketIOMiddleware, self).__call__(environ,
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/middleware.py", line 63, in __call__
App 132017 output:     return self.engineio_app.handle_request(environ, start_response)
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/socketio/server.py", line 604, in handle_request
App 132017 output:     return self.eio.handle_request(environ, start_response)
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/server.py", line 410, in handle_request
App 132017 output:     packets = socket.handle_get_request(
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/socket.py", line 103, in handle_get_request
App 132017 output:     return getattr(self, '_upgrade_' + transport)(environ,
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/socket.py", line 158, in _upgrade_websocket
App 132017 output:     return ws(environ, start_response)
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/async_drivers/threading.py", line 22, in __call__
App 132017 output:     self.ws = Server(environ)
App 132017 output:   File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/simple_websocket/ws.py", line 319, in __init__
App 132017 output:     raise RuntimeError('Cannot obtain socket from WSGI environment.')
App 132017 output: RuntimeError: Cannot obtain socket from WSGI environment.`

**Reply by miguelgrinberg** [reply]
You are using an unsupported deployment platform. As I mentioned before, please review the deployment section of the documentation to learn what the supported options are. CPanel or other shared hosting platforms are unlikely to work.

**Reply by Agufi28** [reply]
> I made some changes, did some debugging, and got a new error. Complete error with traceback `pp 132017 output: [ pid=132017, time=2023-04-24 11:46:24,425 ]: WSGI application raised an exception! App 132017 output: Traceback (most recent call last): App 132017 output: File "/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/wsgi-loader.py", line 199, in main_loop App 132017 output: socket_hijacked = self.process_request(env, input_stream, client) App 132017 output: File "/opt/cpanel/ea-ruby27/root/usr/share/passenger/helper-scripts/wsgi-loader.py", line 333, in process_request App 132017 output: result = self.app(env, start_response) App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/flask/app.py", line 2548, in __call__ App 132017 output: return self.wsgi_app(environ, start_response) App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/flask_socketio/__init__.py", line 43, in __call__ App 132017 output: return super(_SocketIOMiddleware, self).__call__(environ, App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/middleware.py", line 63, in __call__ App 132017 output: return self.engineio_app.handle_request(environ, start_response) App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/socketio/server.py", line 604, in handle_request App 132017 output: return self.eio.handle_request(environ, start_response) App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/server.py", line 410, in handle_request App 132017 output: packets = socket.handle_get_request( App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/socket.py", line 103, in handle_get_request App 132017 output: return getattr(self, '_upgrade_' + transport)(environ, App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/socket.py", line 158, in _upgrade_websocket App 132017 output: return ws(environ, start_response) App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/engineio/async_drivers/threading.py", line 22, in __call__ App 132017 output: self.ws = Server(environ) App 132017 output: File "/home/bumblrin/virtualenv/rcampusbuddy.bumblr.in/3.10/lib/python3.10/site-packages/simple_websocket/ws.py", line 319, in __init__ App 132017 output: raise RuntimeError('Cannot obtain socket from WSGI environment.') App 132017 output: RuntimeError: Cannot obtain socket from WSGI environment.`

Hi, I had the same issue here. In my case it was a misconfigured `SocketIO` object on the python side. I'm using uwsgi with gevent as a webserver. 
The misconfigured constructor was:
``` python
socketio = SocketIO(app, async_mode='gevent')
```
It worked fine after I changed it to:
``` python
socketio = SocketIO(app, async_mode='gevent_uwsgi')
```
I know at the end it wasn't the same problem you had, but it had the same symptoms so I decided to write my solution in case someone else happens to have the same problem. Have a nice day!

**Comment by PrimeTheFirst** [comment]
Update: My hosting provider says: However, for a websocket to work you'd usually need a dedicated port for inbound connections. or something like a proxy-pass configuration. Although, such setup cannot be performed on a Shared Hosting server. I would recommend running either a third-party web-socket server, or consider upgrading to our private server solutions and configuring your application with a proxypass configuration.


What kind of a third-party web-socket server?

**Reply by bluthen** [reply]
I think they mean something that runs on a non-standard port, any of the deployment options listed here: https://flask-socketio.readthedocs.io/en/latest/deployment.html

Maybe you can followup with them what they mean. I suspect anything on their share hosting server will be unsatisfactory though.
